graph/manifest.go

```diff
@@ -8,62 +8,106 @@ import (
 	sig, err := libtrust.ParsePrettySignature(manifestBytes, "signatures")
 	if err != nil {
-		return nil, false, fmt.Errorf("error parsing payload: %s", err)
+		return "", nil, false, fmt.Errorf("error parsing payload: %s", err)
 	}
 
 	keys, err := sig.Verify()
 	if err != nil {
-		return nil, false, fmt.Errorf("error verifying payload: %s", err)
+		return "", nil, false, fmt.Errorf("error verifying payload: %s", err)
 	}
 
 	payload, err := sig.Payload()
 	if err != nil {
-		return nil, false, fmt.Errorf("error retrieving payload: %s", err)
+		return "", nil, false, fmt.Errorf("error retrieving payload: %s", err)
 	}
 
-	var manifestDigest digest.Digest
+	// TODO(stevvooe): It would be a lot better here to build up a stack of
+	// verifiers, then push the bytes one time for signatures and digests, but
+	// the manifests are typically small, so this optimization is not worth
+	// hacking this code without further refactoring.
 
-	if dgst != "" {
-		manifestDigest, err = digest.ParseDigest(dgst)
+	var localDigest digest.Digest
+
+	// verify the local digest, if present in tag
+	if dgst, err := digest.ParseDigest(ref); err != nil {
+		logrus.Debugf("provided manifest reference %q is not a digest: %v", ref, err)
+
+		// we don't have a local digest, since we are working from a tag.
+		// Digest the payload and return that.
+		localDigest, err = digest.FromBytes(payload)
 		if err != nil {
-			return nil, false, fmt.Errorf("invalid manifest digest from registry: %s", err)
+			// near impossible
+			logrus.Errorf("error calculating local digest during tag pull: %v", err)
+			return "", nil, false, err
 		}
-
-		dgstVerifier, err := digest.NewDigestVerifier(manifestDigest)
+	} else {
+		// verify the manifest against local ref
+		verifier, err := digest.NewDigestVerifier(dgst)
 		if err != nil {
-			return nil, false, fmt.Errorf("unable to verify manifest digest from registry: %s", err)
+			// There are not many ways this can go wrong: if it does, its
+			// fatal. Likley, the cause would be poor validation of the
+			// incoming reference.
+			return "", nil, false, fmt.Errorf("error creating verifier for local digest reference %q: %v", dgst, err)
 		}
 
-		dgstVerifier.Write(payload)
+		if _, err := verifier.Write(payload); err != nil {
+			return "", nil, false, fmt.Errorf("error writing payload to local digest reference verifier: %v", err)
+		}
 
-		if !dgstVerifier.Verified() {
-			computedDigest, _ := digest.FromBytes(payload)
-			return nil, false, fmt.Errorf("unable to verify manifest digest: registry has %q, computed %q", manifestDigest, computedDigest)
+		if !verifier.Verified() {
+			return "", nil, false, fmt.Errorf("verification against local digest reference %q failed", dgst)
 		}
+
+		localDigest = dgst
 	}
 
-	if utils.DigestReference(ref) && ref != manifestDigest.String() {
-		return nil, false, fmt.Errorf("mismatching image manifest digest: got %q, expected %q", manifestDigest, ref)
+	// verify against the remote digest, if available
```

Comment:
If we are already verifying against the ref when its a digest, what else does this provide except that the content body is consistent with the digest header?